### PR TITLE
Better Encoders

### PR DESCRIPTION
**Pull request by JIceberg**
# Better Encoders

Please note that we accept pull requests from anyone, but that does not mean it will be merged.

## What kind of change does this PR introduce?
* Feature

## Did this PR introduce a breaking change?
* No

Removes the external encoders to be replaced by the bicc brain functional encoders in the Motor class. Allows the user to specify a distance per pulse (how much distance is covered by a single tick) for cleaner code.

__Please make sure your PR satisfies the requirements of [the contributing page](CONTRIBUTING.md)__

### CODE REVIEW

**Comment by Lunerwalker2** [comment]
bicc ---> bigg

File: examples/src/main/java/com/example/ftclibexamples/DeadWheelsSample.java [hunk 1]
```diff
@@ -34,6 +34,7 @@
 
     public static final double WHEEL_DIAMETER = 2.0;
     public static final double TICKS_PER_REV = 8192;
+    public static final double DISTANCE_PER_PULSE = Math.PI * WHEEL_DIAMETER / TICKS_PER_REV;
```
Comment (Lunerwalker2) [inline examples/src/main/java/com/example/ftclibexamples/DeadWheelsSample.java:37]: We should probably include a term here for any gearing on the dead wheel.

File: examples/src/main/java/com/example/ftclibexamples/SharedOdometry/AutoStaticRobotPose.java [hunk 2]
```diff
@@ -17,17 +15,24 @@
 
     public static final double TRACKWIDTH = 14.31;
     public static final double CENTER_WHEEL_OFFSET = 0.477;
+    public static final double WHEEL_DIAMETER = 2.0;
+    public static final double TICKS_PER_REV = 8192;
+    public static final double DISTANCE_PER_PULSE = Math.PI * WHEEL_DIAMETER / TICKS_PER_REV;
```
Comment (Lunerwalker2) [inline examples/src/main/java/com/example/ftclibexamples/SharedOdometry/AutoStaticRobotPose.java:20]: Gearing term needed

File: examples/src/main/java/com/example/ftclibexamples/SharedOdometry/TeleStaticRobotPose.java [hunk 3]
```diff
@@ -12,17 +11,24 @@
 
     public static final double TRACKWIDTH = 14.31;
     public static final double CENTER_WHEEL_OFFSET = 0.477;
+    public static final double WHEEL_DIAMETER = 2.0;
+    public static final double TICKS_PER_REV = 8192;
+    public static final double DISTANCE_PER_PULSE = Math.PI * WHEEL_DIAMETER / TICKS_PER_REV;
```
Comment (Lunerwalker2) [inline examples/src/main/java/com/example/ftclibexamples/SharedOdometry/TeleStaticRobotPose.java:16]: Gearing term needed

**Review by Lunerwalker2 — CHANGES_REQUESTED** [review]
Everything looks good, however on some of the example we might consider adding a term for gearing just in case a team that is reading those examples needs it. At the very least, even if they are very capable of doing it themselves, it will remind them about it.

**Comment by JIceberg** [comment]
> Everything looks good, however on some of the example we might consider adding a term for gearing just in case a team that is reading those examples needs it. At the very least, even if they are very capable of doing it themselves, it will remind them about it.

Are you sure? That sounds quite specific. Are there really that many teams that gear their odometers in some fashion other than 1:1?

File: examples/src/main/java/com/example/ftclibexamples/DeadWheelsSample.java (hunk 1 above)
Comment (puravdatta-sudo) [reply examples/src/main/java/com/example/ftclibexamples/DeadWheelsSample.java:37]: I think it's pretty self explanatory? With TICKS_PER_REV being adjusted for, however that can def be a comment.

File: examples/src/main/java/com/example/ftclibexamples/InterpLUTSample.java [hunk 4]
```diff
@@ -36,14 +35,22 @@ public void runOpMode() throws InterruptedException
...
-        perpEncoder.setDistancePerPulse(2 / (double)8192);
+        leftEncoder.setDistancePerPulse(2 / (double) 8092);
+        rightEncoder.setDistancePerPulse(2 / (double) 8092);
+        perpEncoder.setDistancePerPulse(2 / (double) 8092);
+
+        odometry = new HolonomicOdometry(
+                leftEncoder::getDistance,
+                rightEncoder::getDistance,
+                perpEncoder::getDistance,
+                14,
```
Comment (puravdatta-sudo) [inline examples/src/main/java/com/example/ftclibexamples/InterpLUTSample.java:51]: Should add comments for this and the following line, not self-explanatory, seems like magic numbers.

File: examples/src/main/java/com/example/ftclibexamples/SharedOdometry/AutoStaticRobotPose.java (hunk 2 above)
Comment (puravdatta-sudo) [reply examples/src/main/java/com/example/ftclibexamples/SharedOdometry/AutoStaticRobotPose.java:20]: Same as above, should be added as a comment or on TICKS_PER_REV, add a comment explaining the process on how to get it.

File: examples/src/main/java/com/example/ftclibexamples/SharedOdometry/TeleStaticRobotPose.java (hunk 3 above)
Comment (puravdatta-sudo) [reply examples/src/main/java/com/example/ftclibexamples/SharedOdometry/TeleStaticRobotPose.java:16]: Same as above.

**Review by puravdatta-sudo — CHANGES_REQUESTED** [review]
Overall, the implementation looks good, I would add a couple comments here and there and add it to the docs and we should be good. As soon as those mentioned comments are added I'll approve it.

**Comment by Lunerwalker2** [comment]
Well, the RR quickstart for example includes it, and there are definitely teams who do it. I like Purav's idea of just adding it in a comment.

**Comment by JIceberg** [comment]
> Well, the RR quickstart for example includes it, and there are definitely teams who do it. I like Purav's idea of just adding it in a comment.

Alright, I'll just add a few comments to the samples instead.

**Comment by NoahBres** [comment]
I have never seen a single team use the gearing term for the encoders :P
Those that do gear (gf design) just gear 1:1.
However, the gearing term is nice for adjusting the wheel radius for fine tuning.

**Review by puravdatta-sudo — APPROVED** [review]
All good, definitely a good addition to FTCLib.

**Comment by NoahBres** [comment]
If you're just gonna put `// if needed, put gearing term here`, just don't bother putting that comment :P
No one is gonna care about the lack of a gear field. Just take out that comment for simplicity. Just gonna be more confusing.

**Review by Lunerwalker2 — APPROVED** [review]
Okay looks good!

**Pull request merged**